### DISCUSSION TITLE
integrating optional DSG question

### DISCUSSION THREAD

**Discussion by 3200pro** [gatsbyjs/gatsby#33958] [Help] [answered]
Hi Everyone,

I'm integrating DSG into pages using a DSG field at Sanity.io. I'm also using the file API routing solution to generate the pages.

Sending it through GraphQL is fine and Gatsby is reading it, however getting the GraphQL query noticed in the DSG Config is really tripping me up. How do you either:

- Add the GraphQL to the config function
- Read the Data in the config function

Here's a working `.tsx` file prior to adding the reference to `pageType`. How would you suggest I add it in. Every process I try says 'Data undefined'

`{sanitySiteRoute.slug__current}.tsx`
```

import React from 'react'
import { graphql } from 'gatsby'
import Layout from "../components/layout"
import Seo from "../components/seo"
  
const Page = ({ data }) => {
  const { title } = data.page
return (
  <Layout>
    <Seo title="Post" />
    <h1>{title}</h1>
    
  </Layout>
)
}

export default Page 

export const query = graphql`
query($id: String) {
  page: sanitySiteRoute(id: { eq: $id }) {
  title
    slug {
      current
    }
    pageType
  }
}
`

export async function config() {
  
  const dsgPosts = true // I'd like to use a ternary here based on pageType === 'DSG'

  return ({ params }) => {
    return {
      defer: dsgPosts
    }
  }
}
```

**Comment by LekoArts** [comment]
Hi!

You would do a query like this:

```graphql
{
  allSanitySiteRoute {
    title
    slug {
      current
    }
    pageType
  }
}
```

And then compare the result of `params.slug__current` with the results from the query.

Here's an example but with markdown, the concept is the same: https://www.gatsbyjs.com/docs/how-to/rendering-options/using-deferred-static-generation/#file-system-route-api

**Reply by 3200pro** [reply]
@LekoArts,

Thanks for the support!

Gave it a shot and Gatsby still isn't capable of mapping the query. Error:  ` TypeError: Cannot read property 'forEach' of undefined`. It seems to be the qraphql that is causing the problem. 

- I logged the params and `slug__current` is correct.
- The query seems to be the problem. I ran the same query in the primary query for the page and everything logs fine there.
- If I comment out `dsgPosts` and set `defer to true` it still errors out with:
` TypeError: Cannot read property 'forEach' of undefined`.
- I am also getting a typescript error for `data` that reads: `Property 'data' does not exist on type 'StaticQueryDocument'.ts(2339)`

```

export async function config() {  
  const { data } = graphql`
    {
      allSanitySiteRoute(filter: {pageType: {eq: "DSG"}}){
        edges {
          node {
            pageType
            slug {
              current
            }
          }
        }
      }
    }
  `
  const dsgPosts = new Set(data.allSanitySiteRoute.edges.node.map(n => n.slug.current))
  return ({ params }) => {
    return {
      defer: dsgPosts.has(params.slug__current)
    }
  }
}
```

**Reply by LekoArts** [reply]
You need to map over edges, not node

**Reply by 3200pro** [reply]
Just gave that a shot. Same error oddly.

`TypeError: Cannot read property 'forEach' of undefined`

`const dsgPosts = new Set(data.allSanitySiteRoute.edges.map(n => n.slug.current))`

**Reply by 3200pro** [reply]
I ran this past a few colleagues and after trying to query site.id it looks like the error may be at a Gatsby level.

**Reply by LekoArts** [reply]
It would need to be `const dsgPosts = new Set(data.allSanitySiteRoute.edges.map(n => n.node.slug.current))`

But if console.log(data) doesn't give you anything then I'd need a minimal reproduction of this to look at. The query should work fine

**Reply by 3200pro** [reply]
Thanks. I sent over an invite to the private repo.

**Reply by LekoArts** [reply]
Thanks! I'll also need the env vars to get this running, you can email me at lennart@gatsbyjs.com

**Reply by 3200pro** [reply]
Sent earlier this morning. Let me know if you didn't get them.

**Comment by LekoArts** [comment, marked as answer]
With this it's all working fine:

![image](https://user-images.githubusercontent.com/16143594/142382852-053cad51-8045-409c-ae81-e2689bd72543.png)

<details>
<summary>Code</summary>

```jsx
import React from 'react'
import { graphql } from 'gatsby'
import Layout from "../components/layout"
import Seo from "../components/seo"
  
const Page = ({ data }) => {
  const { title } = data.page
  return (
    <Layout>
      <Seo title="Post" />
      <h1>{title}</h1>
      
    </Layout>
  )
}

export default Page 

export const query = graphql`
query($id: String) {
  page: sanitySiteRoute(id: { eq: $id }) {
    title
    slug {
      current
    }
    pageType
  }
}
`

export async function config() {
  const { data } = graphql`
    {
      allSanitySiteRoute(filter: {pageType: {eq: "DSG"}}){
        nodes {
          pageType
          slug {
            current
          }
        }
      }
    }
  `

  const dsgPosts = new Set(data.allSanitySiteRoute.nodes.map(node => node.slug.current))

  return ({ params }) => {
    return {
      defer: dsgPosts.has(params.slug__current)
    }
  }
}
```

</details>

The error `TypeError: Cannot read property 'forEach' of undefined` is coming from `gatsby-plugin-typegen` so I commented it out as the plugin needs fixing :)

**Reply by 3200pro** [reply]
Glad to hear its a plugin outside of V4 and thanks for looking through it!
